fsim/upload_owner.go

```diff
@@ -157,13 +193,79 @@ func (u *UploadRequest) finalize() (blockPeer, moduleDone bool, _ error) {
 	if err := u.temp.Close(); err != nil {
 		return false, false, fmt.Errorf("error closing temp file for upload %q: %w", u.Name, err)
 	}
-	// TODO: Enforce chroot-like security
+
+	// Enforce chroot-like security using os.OpenRoot
 	if u.Rename == "" {
 		u.Rename = filepath.Base(u.Name)
 	}
-	oldpath, newpath := u.temp.Name(), filepath.Join(u.Dir, u.Rename)
-	if err := os.Rename(oldpath, newpath); err != nil {
-		return false, false, fmt.Errorf("error renaming temp file %q to %q: %w", oldpath, newpath, err)
+
+	// Create a rooted filesystem to prevent path traversal attacks
+	rootDir, err := os.OpenRoot(u.Dir)
+	if err != nil {
+		return false, false, fmt.Errorf("error creating root filesystem for %q: %w", u.Dir, err)
+	}
+	defer rootDir.Close()
+
+	tempFilePath := u.temp.Name()
+	// Remove the temp file when exiting, ignore errors
+	defer os.Remove(tempFilePath)
+
+	// Check if temp file and destination directory are on the same filesystem
+	samefs, err := sameFilesystem(tempFilePath, u.Dir)
+	if err != nil {
+		return false, false, fmt.Errorf("error checking filesystem for temp file and destination: %w", err)
+	}
+
+	if samefs {
+		// Same filesystem - we can use rename for efficiency
+		// Construct the actual destination path
+		destPath := filepath.Join(u.Dir, u.Rename)
+
+		// Verify the resolved path is still within u.Dir (security check)
```

Comment:
Why is this done rather than using `rootDir` from `os.OpenRoot`?

Reply:
Refactored everything, I hope everything looks good now